torch/testing/_internal/common_methods_invocations.py

```diff
@@ -349,6 +349,70 @@ def wrapped_fn(x):
 
     return wrapped_fn
 
+def np_binary_ufunc_type_promotion_wrapper(fn):
+    # Wrapper that passes PyTorch's default scalar
+    #   type as an argument to the wrapped NumPy
+    #   binary ufunc when given an integer input .
+    #   This mimicks PyTorch's integer->floating point
+    #   type promotion.
+    #
+    # This is necessary when NumPy promotes
+    #   integer types to double, since PyTorch promotes
+    #   integer types to the default scalar type.
+    #
+    # Besides, the table below shows the difference between
+    #   PyTorch and NumPy on the type promotion for binary
+    #   functions
+    #      inputs         NumPy     PyTorch
+    #    bool  bool       vary        bool
+    #   int16 float16    float32    float16
+    #   int32 float16    float64    float16
+    #   int64 float16    float64    float16
+    #   int32 float32    float64    float32
+    #   int64 float32    float64    float32
+    #
+    # Another special case only on Windows platform due to
+    # torch tensor wrapper for integer scalar is int64, but
+    # numpy wrapper for integer scalar is int32.
+    # For boolean tensor and integer scalar, NumPy result
+    # needs to be promoted from int32 to int64
+
+    # Helper to get the unified type between PyTorch
+    #    and NumPy
+    #    x is Tensor; y can be Tensor or Scalar
+    def unified_type(x, y):
+        if not hasattr(y, 'dtype'):
+            if x.dtype in [np.bool, np.uint8, np.int8, np.int16, np.int32, np.int64] and \
+                    type(y) == float:
+                return torch_to_numpy_dtype_dict[torch.get_default_dtype()]
+            if IS_WINDOWS and x.dtype == np.bool and type(y) == int:
+                return np.int64
+            return None
+        if x.dtype == np.bool and y.dtype == np.bool:
+            return np.bool
+        if (x.dtype == np.float16 and y.dtype in [np.int16, np.int32, np.int64]) or \
+           (x.dtype in [np.int16, np.int32, np.int64] and y.dtype == np.float16):
+            return np.float16
+        if (x.dtype == np.float32 and y.dtype in [np.int32, np.int64]) or \
+           (x.dtype in [np.int32, np.int64] and y.dtype == np.float32):
+            return np.float32
+        return None
```

Comment:
Facts of binary operations:
1. First argument must be Tensor
2. Second argument can be Tensor or Scalar

There are several cases with different type promotion between NumPy and PyTorch for binary operations:
1. Tensor & Scalar
    - Integer Tensor & Float Scalar
      - np: np.float64
      - torch: torch.float32
    - Boolean Tensor & Integer Scalar on Windows
      - Due to scalar becomes np.int64 when converted to numpy array, but torch.int32 when converted to Tensor only on Windows
      - np: np.int32
      - torch: torch.int64
2. Tensor & Tensor
    - Boolean Tensor & Boolean Tensor
      - np: different based on platform
      - torch: torch.bool
    - Integral Tensor & Float Tensor / Float Tensor & Integral Tensor
      - Shown in the table in comment

Reply:
Add this wrapper for all binary operations required to comparing the results between NumPy and PyTorch

Reply:
can you clarify what `np: different based on platform` means?

Also, perhaps split out the bool changes since they seem more complicated than the rest?